File: wiki-java/src/main/java/dev/freya02/wiki/localization/LocalLocalizationMapReader.java
package dev.freya02.wiki.localization;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.freya022.botcommands.api.core.BContext;
import io.github.freya022.botcommands.api.core.Logging;
import io.github.freya022.botcommands.api.localization.LocalizationMapRequest;
import io.github.freya022.botcommands.api.localization.readers.AbstractJacksonLocalizationMapReader;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;
import org.slf4j.Logger;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;

// --8<-- [start:local_localization_map_reader-java]
public class LocalLocalizationMapReader extends AbstractJacksonLocalizationMapReader {
    private static final Logger LOGGER = Logging.getLogger();

    private final Path basePath;

    public LocalLocalizationMapReader(@NonNull BContext context, @NonNull ObjectMapper mapper, @NonNull Path basePath) {
        super(context, mapper);
        this.basePath = basePath;
    }

    @Nullable
    @Override
    public InputStream getInputStream(@NonNull LocalizationMapRequest request) {
        var fileName = basePath + ".json";
        var path = basePath.resolve(fileName);
        if (Files.notExists(path)) {
            LOGGER.trace("Found no bundle at {}", path);
            return null;
        }

        try {
            return Files.newInputStream(path);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}
// --8<-- [end:local_localization_map_reader-java]

File: wiki-kotlin/src/main/kotlin/io/github/freya022/wiki/localization/LocalLocalizationMapReader.kt
package io.github.freya022.wiki.localization

import com.fasterxml.jackson.databind.ObjectMapper
import io.github.freya022.botcommands.api.core.BContext
import io.github.freya022.botcommands.api.localization.LocalizationMapRequest
import io.github.freya022.botcommands.api.localization.readers.AbstractJacksonLocalizationMapReader
import io.github.oshai.kotlinlogging.KotlinLogging
import java.io.InputStream
import java.nio.file.Path
import kotlin.io.path.inputStream
import kotlin.io.path.notExists
import kotlin.io.path.pathString

private val logger = KotlinLogging.logger { }

// --8<-- [start:local_localization_map_reader-kotlin]
class LocalLocalizationMapReader(
    context: BContext,
    mapper: ObjectMapper,
    private val basePath: Path,
) : AbstractJacksonLocalizationMapReader(context, mapper) {

    override fun getInputStream(request: LocalizationMapRequest): InputStream? {
        val fileName = "${request.bundleName}.json"
        val path = basePath.resolve(fileName)
        if (path.notExists()) {
            logger.trace { "Found no bundle at ${path.pathString}" }
            return null
        }

        return path.inputStream()
    }
}
// --8<-- [end:local_localization_map_reader-kotlin]
